### DISCUSSION TITLE
[C128] #pragma wrapped-call with dynamic banking?

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#1938] [Q&A] [answered]
Currently I make extensive use of the ‘wrapping’ #pragma which works very smoothly.
Especially now I have the multiple parameters working as well. 

There is however still something to desire;

Currently I have to define the ‘same’ function multiple times with just different names and different #pragma's.

Just because they are used in different banking (RAM#n) configurations like; 
            #pragma wrapped-call (push, bankswitch, 0x4e) // RAM #1 Kernal and I/O only
                void POKE_FAR_RAM_1 (int address, char value);
            #pragma wrapped-call (pop)

            #pragma wrapped-call (push, bankswitch, 0x8e) // RAM #2 Kernal and I/O only
                 void POKE_FAR_RAM_2(int address, char value);
            #pragma wrapped-call (pop)

However, is there clever way in which it is possible to consolidate that in one way or the other? 

What would be nice is to have a parameter in the function which just indicates the bank/(ram cfg) the function should be executed in.
Where the function can be defined in the COMMONRAM; 

          #pragma code-name (push, “COMMONRAM”) 
                void POKE_FAR (char bankcfg, int address, char value);
          #pragma code-name (pop) 

         #pragma wrapped-call (push, bankswitch, bankcfg) // RAM #1, #2 or #3
                 void POKE_FAR(char bankcfg, int address, char value);
            #pragma wrapped-call (pop)

But that would require to have the bank/(ram cfg) indicator in the wrapped-call to be dynamic instead of static.
(I know there is the 'bank' value which can be used in that place but that is not doing the trick needed as that is considered 'static' Can only be one predefined value which is put at the c128.cfg file at a certain RAM#x
 
Perhaps there is a way to manipulate/overwrite the used ‘tmp4’ directly based on the ‘bankcfg’ variable (in the function) directly perhaps? (despite this would be an ugly solution). 

Alternatively, within the ‘bankswitch.s routine’ I could add an extra ‘value lookup’ which checks the actual ram configuration (bank)  to execute the function in coming from the value in the function? 

**Question #1:** Is there perhaps a simple way I am overlooking which would make this possible? 
**Question #2**: What would be needed on the cc65 toolchain to make this possible? (could this be solved in the #pragma wrapped-call?

The above is just for possible optimization as it all does work nice but with some unneeded redundancies.

**Comment by mrdudz** [comment]
I have to admit i don't quite understand what you are trying to do there :)

**Reply by RetrofanC** [reply]
Sorry for not being clear, let me try to be more clear. 
(I might tell things which are obvious for you but I need it it to tell the story) 

In the C128 you select RAM (0-3) and RAM configurations (with kernel, with kernel & I/O or only full RAM etc.) 

So I want to be able to define only one function; 

void POKE_FAR (char bankcfg, int address, char value) 

then i want to be able to use it as follows; 

POKE_FAR (0x4E, 0xD020, 0x01);  // RAM#1 -  Config Kernel & I/O -- Border will change color 
POKE_FAR (0x7F, 0xE000, 0x02);  //  RAM#1 -  Config only RAM -- Write on 0xE000 which is 'under' the Kernel in RAM#1 
POKE_FAR (0x3F, 0xE000, 0xDE);  // RAM#0 -  Config only RAM -- Write on 0xE000 which is 'under' the Kernel in RAM#0 (16k extra!) 
POKE_FAR (0x8E, 0x4000, 0xDE);  // RAM#2  - Config  only RAM 

So the 'bankcfg' parameter will determine in which the function needs to run and what the configuration of that ram environment is. 

This is what I have when I am using my own .asm rountine. 
But that takes up another extra range of memory in the common ram area.
So I want to be able to use the #pragma wrapped-call or something standard to make this work as well.

Does the above description makes it more clear?

**Comment by mrdudz** [comment, marked as answer]
> So the 'bankcfg' parameter will determine in which the function needs to run and what the configuration of that ram environment is.

That sounds weird.... you will still need a trampoline function in common ram, and then a copy of the actual function in every bank/memory config its supposed to run in.

In any case, it cant be done with the wrapped call - you'll have to roll your own :)
